core-rust/state-manager/src/store/gc.rs

```diff
@@ -0,0 +1,459 @@
+        all_deleted
+    }
+}
+
+struct DfsIterator<'s, S> {
+    tree_store: &'s S,
+    levels: Vec<VecDeque<(NodeKey, bool)>>,
+}
+
+impl<'s, S: ReadableTreeStore> DfsIterator<'s, S> {
+    pub fn new(tree_store: &'s S, root_key: &NodeKey) -> Self {
+        let levels = Self::drill_levels(tree_store, root_key);
+        Self { tree_store, levels }
+    }
+
+    fn drill_levels(tree_store: &S, start_key: &NodeKey) -> Vec<VecDeque<(NodeKey, bool)>> {
+        let mut levels = Vec::new();
+        let mut at_key = start_key;
+        // Drill down the leftmost descendants chain, in order to:
+        // - find the starting point of the DFS iteration;
+        // - and record the single chain of "siblings of ancestors" (to easily continue the traversal).
+        loop {
+            let Some(at_node) = tree_store.get_node(at_key) else {
+                break; // let's silently tolerate nodes deleted by any previous incomplete GC run
+            };
+            match at_node {
+                TreeNode::Internal(internal) => {
+                    let level = internal
+                        .children
+                        .into_iter()
+                        .map(|entry| {
+                            (
+                                at_key.gen_child_node_key(entry.version, entry.nibble),
+                                entry.is_leaf,
+                            )
+                        })
+                        .collect::<VecDeque<_>>();
+                    levels.push(level);
+                    let (leftmost_child_key, is_leaf) = levels
+                        .last()
+                        .expect("we literally just pushed an element into it")
+                        .front()
+                        .expect("if internal node exists, then it has children");
+                    if *is_leaf {
+                        break;
+                    } else {
+                        at_key = leftmost_child_key;
+                    }
+                }
+                TreeNode::Leaf(_) => {
+                    // This special case may only happen on the initialization (i.e. from `new()`);
+                    // As seen in the branch above, we normally do not even attempt to load the
+                    // final leaf node from DB.
+                    levels.push(VecDeque::from([(at_key.clone(), true)]));
+                    break;
+                }
+                TreeNode::Null => {
+                    // This case does not occur in practice at all (we do not keep empty lower-tier
+                    // trees in the DB - we remove their higher-tier leaf node instead).
+                    break;
+                }
+            }
+        }
+        levels
+    }
+}
+
+impl<'s, S: ReadableTreeStore> Iterator for DfsIterator<'s, S> {
+    type Item = NodeKey;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Some(mut current_deepest_level) = self.levels.pop() else {
```

Comment:
Personally, If starting from the beginning with this, I'd store the structure as:

`current_path: Vec<Rc<NodeData>>`

```rust
struct NodeData {
    node_key: NodeKey,
    state: IterationState; // Starts as `IsLeaf` or `NotYetLoaded`
}

impl NodeData {
  pub new_with_unknown_status(node_key: NodeKey) -> Self {
    Self { node_key, state: IterationState::NotYetLoaded }
  }

  pub new_for_leaf_assuming_it_exists(node_key: NodeKey) -> Self {
    Self { node_key, state: IterationState::Children(None), }
  }
}

enum IterationState {
    NotYetLoaded,
    Deleted,
    Children(Option<Vec<Rc<NodeData>>>),
    Visited,
}
```

And then I'd start with `self.current_path: vec![Rc::new(NodeData::new_with_unknown_status(root))]` and then in the iterator itself, I loop by looking at last entry in `current_path`:
* If its state is `ChildrenNotYetLoaded` I load it from the DB, and set it to either `Deleted` or `ChildrenRemaining(X)` (for a leaf, X = empty).
* If it's `Deleted` or `Visited`, pop it off `current_path`
* At this point, its state is ChildrenRemaining, so iterate over these, ignoring any which are `Visited` or `Deleted`. Then:
  * If is is empty, set the current node to `Visited`, pop it off `current_path`, and return its `node_key`
  * If it is not empty, push the first child to the end of `current_path`.

And then - if you want to optimize to minimize writes, but potentially return nodes that don't exist, you can have `ChildrenRemaining` leaves create straight as `NodeData::new_for_leaf_assuming_it_exists` instead of `NodeData::new_with_unknown_status`. OR you just use `NodeData::new_with_unknown_status`, and force a lookup to see if the node already exists before returning it.

Reply:
I don't have good news...

I have rewritten the iterator into the above impl.
And then I simplified the `NotYetLoaded` state out.
And then I noticed that loading anything other than `InternalNode` from DB (e.g. a leaf) is a special case that should really be extracted out to a pre-processing logic (before returning the iterator).
And then 90% of what was left could really be rewritten into a single `.flat_map()`.
And then I realized that the entire logic of a DFS post-order lazy iterator can be simply expressed by its recurrent definition ("iterate children recursively, and then add parent") and that in FP language it is `return parent.children.flat_map(...).chain(parent);`.

So, that's what I now pushed: essentially 10 lines of code + 20 lines of comments, PTAL.

Reply:
⚠️ The issue with removing the `NotYetLoaded` state as written is that the algorithm becomes greedy and you load the whole tree into memory immediately.

Reply:
> algorithm becomes greedy and you load the whole tree into memory immediately

I don't believe `.flat_map()` behaves like this...?

Reply:
Just reviewing the current code. Yes, with `flat_map` it looks good I think...

You've essentially moved all the iterator "state" into the recursive stack.